components/compliance-service/inspec-agent/runner/runner.go

```diff
@@ -0,0 +1,712 @@
+	if len(tc.SecretsArr) > 0 {
+		return true
+	}
+	if len(tc.User) == 0 {
+		return false
+	}
+	if len(tc.Password) == 0 {
+		if len(tc.KeyFiles) == 0 {
+			return false
+		}
+	}
+	return true
+}
+
+func cloudEnvVars(tc *inspec.TargetConfig) (map[string]string, error) {
+	envsMap := map[string]string{
+		"CHEF_LICENSE": "accept-no-persist",
+	}
+	switch tc.Backend {
+	case "aws":
+		if tc.AwsUser == "" || tc.AwsPassword == "" {
+			logrus.Debugf("no aws creds found in env vars, no aws creds found for node; attempting to use aws credential chain via inspec/train")
+			return envsMap, nil
+		}
+		envsMap["AWS_ACCESS_KEY_ID"] = tc.AwsUser
+		envsMap["AWS_SECRET_ACCESS_KEY"] = tc.AwsPassword
+
+		// Only set a TOKEN ENV variable when one is needed.
+		// Otherwise it prevents the TOKEN-less account credentials from working
+		if tc.AwsSessionToken != "" {
+			envsMap["AWS_SESSION_TOKEN"] = tc.AwsSessionToken
+		}
+		return envsMap, nil
+	case "azure":
+		if tc.AzureClientID == "" || tc.AzureClientSecret == "" || tc.AzureTenantID == "" {
+			logrus.Debugf("no azure creds found in environment, no azure creds found for node; attempting to use azure credential chain via inspec/train")
+			return envsMap, nil
+		}
+		envsMap["AZURE_CLIENT_ID"] = tc.AzureClientID
+		envsMap["AZURE_CLIENT_SECRET"] = tc.AzureClientSecret
+		envsMap["AZURE_TENANT_ID"] = tc.AzureTenantID
+		return envsMap, nil
+	case "gcp":
+		if tc.GcpCredsJson != "" {
+			// Specify "" for the temp dir as ioutil will pick TMPDIR or OS default
+			tmpFile, err := ioutil.TempFile("", ".gcp-project-cred.json")
+			if err != nil {
+				return envsMap, err
+			}
+
+			err = ioutil.WriteFile(tmpFile.Name(), []byte(tc.GcpCredsJson), 0400)
+			if err != nil {
+				return envsMap, err
+			}
+
+			// Not consumed by InSpec via the json config stdin but via file on disk
+			tc.GcpCredsJson = ""
+			envsMap["GOOGLE_APPLICATION_CREDENTIALS"] = tmpFile.Name()
+			return envsMap, nil
+		} else {
+			return envsMap, fmt.Errorf("cloudEnvVars: GcpCredsJson can't be empty, job will fail execution")
+		}
+	}
+
+	return envsMap, nil
+}
+
+// doDetect executes a detect job and returns error type for retrying purposes
+func doDetect(job *types.InspecJob) (osInfo *inspec.OSInfo, err *inspec.Error) {
+	timeout := time.Duration(job.Timeout) * time.Second
+	env, genericErr := cloudEnvVars(&job.TargetConfig)
+	defer func() {
+		cleanupCreds(env)
+	}()
+	if genericErr != nil {
+		return nil, inspec.NewInspecError(inspec.UNKNOWN_ERROR, genericErr.Error())
+	}
+
+	for i, tc := range potentialTargetConfigs(job) {
+		osInfo, err = inspec.Detect(&tc, timeout, env)
+		if err == nil {
+			break
+		}
+		logrus.Errorf("%s(%s) connection attempt # %d failed for node(%s) with error: %s", job.JobType, job.JobID, i+1, job.NodeID, err.Message)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	job.NodeStatus = types.StatusCompleted
+	return osInfo, nil
+}
+
+func doExec(job *types.InspecJob) (jsonBytes []byte, err *inspec.Error) {
```

Comment:
this is a question for the compliance folks: do we want to wire through cancellation? it looks like the run function in inspec/cli.go already uses a context, so it would require threading our context to that function. Doing so would mean that when cancelling a workflow, all the inspec processes get killed.

Reply:
"do we want to wire through cancellation" -- can you define what you mean by cancellation? like, as a user, how does one "cancel"?

Reply:
I think there are 2 parts.
1.) how does a user cancel
2.) Once a user cancels, what do we do. Assuming a user had some way to cancel, currently running scan scans (inspec is running) would wait to finish. If we thread the context to the doExec function, we can kill the inspec process when the user cancels as the context will automatically get canceled in that case